### PR TITLE
Initial framework for dns tunneling

### PR DESCRIPTION
**Pull request by uruwhy**
## Description
Initial framework for DNS tunneling contact as gocat extension. Does not yet support file uploads.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?
Ran operations with agents on windows, linux, and Mac. tested abilities that downloaded payloads and submitted results back to the C2.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: gocat-extensions/contact/dns_tunneling.go [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+	}
+	return base64.StdEncoding.DecodeString(buffer.String())
+}
+
+func generateQname(messageID string, messageType string, chunkIndex int, numChunks int, data []byte) (string, error) {
+	if len(data) > MAX_UPLOAD_CHUNK_SIZE {
+		return "", errors.New("Data chunk too large.")
+	}
+	dataHex := hex.EncodeToString(data)
+	//return fmt.Sprintf("%s.%s.%d.%d.%s.%s", messageID, messageType, chunkIndex, numChunks, dataHex, BASE_DOMAIN), nil
```
Comment (mrengstrom) [inline gocat-extensions/contact/dns_tunneling.go:266]: is the comment needed?  Seems like it can be dropped.

Comment (uruwhy) [reply gocat-extensions/contact/dns_tunneling.go:266]: good eye, can't remember why I left that in there. I'll clean up some other vestigial comments that are hanging around as well

Comment (uruwhy) [reply gocat-extensions/contact/dns_tunneling.go:266]: Ah, I think the DNS library I used for golang required qnames to end with a ".", hence line 267. Regardless, I'll remove the old comment

**Review by wbooth — CHANGES_REQUESTED** [review]
.

File: gocat-extensions/contact/dns_tunneling.go (hunk 1 above)
Comment (uruwhy) [reply gocat-extensions/contact/dns_tunneling.go]: fixed

**Review by mrengstrom — APPROVED** [review]

**Pull request merged**